### PR TITLE
Add reviewer and requester roles.

### LINKED ISSUE #27321: Create `reviewer` and `requester` preset roles
Success criteria:

- Teleport provides “reviewer” preset role that allows review access to all groups and apps.
- Teleport provides "requester" preset role that allows users to request access to apps and groups via “access” and “access_groups” presets we already have.
- Getting started guide is updated to use “reviewer” and “editor” roles for admins, and as “access” and “requester” for regular users.

### PR DESCRIPTION
**Pull request by mdwn**
Reviewer and requester roles have been added to allow for easy defaults for reviewing and requesting applications and user groups for the Okta service.

Closes https://github.com/gravitational/teleport/issues/27321

### CODE REVIEW

File: constants.go
Comment (zmb3) [inline constants.go:621]:
```suggestion
	// for requesting access to resources.
```

Though if these roles are only for Okta stuff, we should give them more descriptive names and godocs.

**Review by nklaassen — COMMENTED** [review]
Not the end of the world, but will this break guides that tell users to create roles named requester and reviewer? eg https://goteleport.com/docs/access-controls/access-requests/resource-requests/#step-18-create-the-requester-role ninja-edit: Oh I see your latest commit just updated all those guides

Actually, what happens if you already have a role with one of those names, and upgrade to this version? I guess nothing would change unless those roles named `requester` or `reviewer` didn't have any `allow.request` or `allow.review` permissions, which seems fairly unlikely. We'd definitely want to release note this though

I wonder if there's a good way to scope these role names more tightly so it's clear they can't request/review *anything*, just some specific things. Unless we *do* want to make these roles able to request/review everything requestable/reviewable. But I assume there is some motivation as to why we want preset roles for this that is informing that decision

File: docs/pages/access-controls/access-requests/resource-requests.mdx
Comment (zmb3) [inline docs/pages/access-controls/access-requests/resource-requests.mdx:42]: Should we instead just ask the user to assign the new preset `requestor` role rather than create another one?

File: lib/services/presets.go
Comment (zmb3) [inline lib/services/presets.go:225]: What's this for?

File: constants.go
Comment (mdwn) [reply constants.go:621]: The use case is primarily Okta, but they're more general than that.

File: docs/pages/access-controls/access-requests/resource-requests.mdx
Comment (mdwn) [reply docs/pages/access-controls/access-requests/resource-requests.mdx:42]: I was wondering about that. I think there's some use in having this walk through of creating review/request roles from scratch, so I decided to leave this.

File: lib/services/presets.go
Comment (mdwn) [reply lib/services/presets.go:225]: I added this (basically coping off `auditor`) to ensure that the reviewer/requester role can't log into anything on its own. In other words, it's purely for reviewing/requesting and nothing else.

Comment (zmb3) [reply lib/services/presets.go:225]: Wouldn't omitting logins completely achieve the same result?

File: docs/pages/access-controls/access-requests/resource-requests.mdx
Comment (zmb3) [reply docs/pages/access-controls/access-requests/resource-requests.mdx:42]: I dunno, I think there are plenty of other places we can teach a user how to create a role from scratch. The goal of this guide is to see access requests in action. If your new changes allow us to get the user to that outcome faster, I think we should do it.

File: lib/services/presets.go
Comment (r0mant) [inline lib/services/presets.go:247]: Why do we need these options on the requester/reviewer roles?

Comment (mdwn) [reply lib/services/presets.go:247]: Bad copy paste. :-X

**Comment by mdwn** [comment]
Wanted to ping on this @r0mant, @zmb3, @nklaassen. Any other thoughts here?

File: lib/services/presets.go
Comment (r0mant) [inline lib/services/presets.go:224]: Hmm now that I'm thinking about this, having "empty" requester and reviewer roles in OSS would be pretty confusing. I think we should do this:

- In OSS do not create these presets at all instead of creating empty roles as they're pretty useless without allow clauses.
- In the docs, only add "requester" and "reviewer" roles when Enterprise scope is selected.

**Review by r0mant — COMMENTED** [review]
Also, can you make sure that if a user already has existing requester/reviewer roles in Enterprise, we don't overwrite those?

**Comment by ptgott** [comment]
@mdwn Looks like there are some engineering questions to work out. Do these affect the docs aspect? If so, let me know when the implementation is done and I'll give this a docs review!

File: lib/services/presets.go
Comment (mdwn) [reply lib/services/presets.go]: Okay, this has been done (save any docs syntax issues).

**Review by r0mant — APPROVED** [review]
lgtm if you've verified customer-created roles with the same names aren't being overridden.

**Comment by mdwn** [comment]
> lgtm if you've verified customer-created roles with the same names aren't being overridden.

I'll double check before merging

**Comment by tcsc** [comment]
Forgive the possibly boneheaded question here, as this is my first time diving into Teleport RBAC in a _long_ time:  The new `reviewer` role is only allowed to grant access to the `access` and `group_access` roles, yes?

I originally misread it as being able to globally approve access to _any_ role, which I don't think is what's intended.

**Review by r0mant — CHANGES_REQUESTED** [review]
Requesting changes until we make sure we don't change customer resources if they happen to have the same name.

**Comment by mdwn** [comment]
> Requesting changes until we make sure we don't change customer resources if they happen to have the same name.

I've adjusted this so that customer roles will not be overwritten.

File: lib/services/presets.go
Comment (nklaassen) [inline lib/services/presets.go:265]: does it allow requesting and reviewing?

File: lib/services/presets.go
Comment (nklaassen) [inline lib/services/presets.go:419]: I don't think we should blindly overwrite this. Imagine someone sets it to false, we should probably respect that and not touch the role. Probably we should only set it on creation, or set it if it's unset for OG preset roles (editor,access,auditor)

File: lib/services/presets.go
Comment (nklaassen) [inline lib/services/presets.go:301]: this is semantically different than the other `defaultXXX` functions, maybe it needs another name like `bootstrapRoleLabels`

File: lib/services/presets_test.go
Comment (nklaassen) [inline lib/services/presets_test.go:57]: I don't think that we should assert that the labels change when we return an error and expect the role not to be written. Maybe this should actually trigger a change to write the new label

File: lib/services/presets_test.go
Comment (nklaassen) [inline lib/services/presets_test.go:75]: for the existing roles, can we add test cases with/without the managed label already set?

File: api/types/constants.go
Comment (tcsc) [inline api/types/constants.go:571]: I'm not 100% clear on what the implications of _managed by Teleport_ and _These resources can be freely modified by Teleport if needed_ are in this context. 

I'm reading it as:
> DANGER! This resource should not be modified bay anyone who is _not_ a Teleport auth server.  Teleport may overwrite/revert/otherwise destroy any changes to this resource at will.

Is that about right? If so, at some point we should add some flashing lights in the UI if you try to edit it.

File: docs/pages/deploy-a-cluster/helm-deployments/custom.mdx
Comment (mdwn) [reply docs/pages/deploy-a-cluster/helm-deployments/custom.mdx:219]: Ahhh, this was a bad copy paste. I'll revert this.

File: lib/services/presets.go
Comment (mdwn) [reply lib/services/presets.go:265]: It doesn't allow it, but it supports it -- it's used by reviewer and requester roles for enabling user group requests and user group reviews.

File: lib/services/presets.go
Comment (mdwn) [reply lib/services/presets.go:373]: I'm not sure about this one. I don't know that we want to grant those roles explicitly, just resources that belong to those roles.

File: docs/pages/access-controls/guides/impersonation.mdx
Comment (mdwn) [reply docs/pages/access-controls/guides/impersonation.mdx:120]: Pretty much all of the references add in the doc here are to add visibility for the reviewer/requester roles. I'm amenable to backpedaling on that, but want to make sure we're on the same page (cc @r0mant)

File: docs/pages/application-access/getting-started.mdx
Comment (mdwn) [reply docs/pages/application-access/getting-started.mdx:196]: You could very well need it for user group requests, which are not granted as part of access.

File: docs/pages/access-controls/sso/github-sso.mdx
Comment (mdwn) [reply docs/pages/access-controls/sso/github-sso.mdx:273]: I'll fix up that `teams_to_roles` map and remove the `-teleport-internal-join`.

File: api/types/constants.go
Comment (mdwn) [reply api/types/constants.go]: I've updated the wording here -- do you mind taking a look when you get a chance?

File: docs/pages/access-controls/guides/impersonation.mdx
Comment (mdwn) [reply docs/pages/access-controls/guides/impersonation.mdx:120]: Roman is fine with backpedaling on irrelevant references to the reviewer/requester roles, so I'll revert the changes where you've mentioned.

File: lib/services/presets.go
Comment (nklaassen) [reply lib/services/presets.go:232]: @mdwn can we get rid of this?

File: lib/services/presets.go
Comment (nklaassen) [inline lib/services/presets.go:260]: if we resolve the discussion in https://github.com/gravitational/teleport/pull/27673/files#r1224472116 can we remove this here and the the group-access role? (gh won't let me comment on the group-access)

```suggestion
```

File: lib/services/presets.go
Comment (nklaassen) [reply lib/services/presets.go:373]: I guess that's more of a product/ux question, but I think it will be hard to change our minds on this once we ship it, I won't block it either way but I think it should be given some thought

File: lib/services/presets.go
Comment (nklaassen) [inline lib/services/presets.go:412]: nit:
```suggestion
			metadata.Labels = make(map[string]string, len(defaultRoleLabels))
```

File: docs/pages/access-controls/getting-started.mdx
Comment (ptgott) [inline docs/pages/access-controls/getting-started.mdx:29]: Users tend not to notice that we have scope switchers at the top of docs pages. We should use something besides `ScopedBlock` here, since this is information that's important not to miss. Maybe a `Details` box?

File: docs/pages/access-controls/getting-started.mdx
Comment (ptgott) [inline docs/pages/access-controls/getting-started.mdx:30]:
```suggestion
Teleport Enterprise contains two additional preset roles: `reviewer` and `requester`.
```

File: docs/pages/access-controls/getting-started.mdx
Comment (ptgott) [inline docs/pages/access-controls/getting-started.mdx:32]: I'm not sure the "application and user group" terminology is something we use in the rest of the docs. Maybe "resources and roles" instead? Also, other parts of the docs capitalize Access Requests.

File: docs/pages/access-controls/reference.mdx
Comment (ptgott) [inline docs/pages/access-controls/reference.mdx:131]:
```suggestion
| `requester`| Enterprise-only role that allows a user to create Access Requests. |
| `reviewer`| Enterprise-only role that allows review of Access Requests. |
```

File: docs/pages/application-access/getting-started.mdx
Comment (ptgott) [inline docs/pages/application-access/getting-started.mdx:189]: If it's important that Enterprise users not miss this part of the page, we should put this in a `Tabs` rather than a `ScopedBlock`, since users tend not to realize you can manipulate the scope switcher at the top of the page.

File: lib/services/presets_test.go
Comment (nklaassen) [reply lib/services/presets_test.go:60]: No I misread, I thought this was `expected:`

File: docs/pages/get-started.mdx
Comment (ptgott) [inline docs/pages/get-started.mdx:109]: This page is intended only for OSS users, so I don't think we need the `ScopedBlock` here

File: docs/pages/connect-your-client/introduction.mdx
Comment (nklaassen) [inline docs/pages/connect-your-client/introduction.mdx:41]:
```suggestion
$ tsh login --proxy=<Var name="teleport.example.com"/> --auth=github
```

File: docs/pages/application-access/getting-started.mdx
Comment (nklaassen) [inline docs/pages/application-access/getting-started.mdx:182]:
```suggestion
Next, let's create a user to access the application we've just connected. Teleport has a built-in role called `access` that allows users to access cluster resources. Create a local user assigned to this role:
```

File: docs/pages/application-access/getting-started.mdx
Comment (nklaassen) [inline docs/pages/application-access/getting-started.mdx:191]: if I read "`access` that allows users to access cluster resources" + "`requester` that allows users to request resources" as a normal teleport user I'm going to be really confused (actually, I am a bit confused myself). If the `requester` role is relevant to this guide can we say why it's actually needed?

File: docs/pages/application-access/getting-started.mdx
Comment (mdwn) [reply docs/pages/application-access/getting-started.mdx:189]: I don't think it's super dire if they miss this.

File: docs/pages/application-access/getting-started.mdx
Comment (mdwn) [reply docs/pages/application-access/getting-started.mdx:191]: I think I'll remove it from this guide, might makes things simpler.

File: lib/services/presets.go
Comment (mdwn) [reply lib/services/presets.go]: Okay, I've removed this and from the other new roles. I've left the auditor role intact, however.

File: api/types/constants.go
Comment (r0mant) [inline api/types/constants.go:573]: @tcsc also introduced a "system resource" label in https://github.com/gravitational/teleport/pull/27829. Let's please not have 2 different labels for the same thing. I'd probably vote for "teleport.internal/system: true".

Comment (mdwn) [reply api/types/constants.go]: I added in the labels as we discussed. Let me know what you think.

File: api/types/constants.go
Comment (r0mant) [inline api/types/constants.go:584]: Nit: For consistency with `PresetResource`.
```suggestion
	SystemResource = "system"
```

File: lib/services/presets.go
Comment (r0mant) [inline lib/services/presets.go:428]: Just making sure: will caller handle this error gracefully and not crash the startup?

File: lib/services/presets.go
Comment (r0mant) [reply lib/services/presets.go:370]: TBH I think we can always relax the role and add more permissions to it (such as requesting roles in addition to resources) if we decide it improves the UX.

**Review by r0mant — APPROVED** [review]

File: lib/services/presets.go
Comment (mdwn) [reply lib/services/presets.go:428]: That's correct.

**Review by nklaassen — APPROVED** [review]

**Comment by mdwn** [comment]
I'm going to go ahead and merge this/backport it and if there are any docs concerns, we can handle them in a follow up.

**Pull request merged**
